src/ccl_correlation.c

```diff
@@ -194,81 +193,103 @@ static double corr_bessel_integrand(double l,void *params)
-    return;
-  }
-  switch(corr_type) {
-  case CCL_CORR_GG :
-    cp->i_bessel=0;
-    break;
-  case CCL_CORR_GL :
-    cp->i_bessel=2;
-    break;
-  case CCL_CORR_LP :
-    cp->i_bessel=0;
-    break;
-  case CCL_CORR_LM :
-    cp->i_bessel=4;
-    break;
-  }
-
-  if(cls[0]*cls[1]<=0)
-    cp->extrapol_0=0;
-  else {
-    cp->extrapol_0=1;
-    cp->tilt0=log10(cls[1]/cls[0])/log10(ell[1]/ell[0]);
-  }
-
-  if(cls[n_ell-2]*cls[n_ell-1]<=0)
-    cp->extrapol_f=0;
-  else {
-    cp->extrapol_f=1;
-    cp->tiltf=log10(cls[n_ell-1]/cls[n_ell-2])/log10(ell[n_ell-1]/ell[n_ell-2]);
-  }
-
   int ith, gslstatus;
   double result,eresult;
   gsl_function F;
-  gsl_integration_workspace *w=gsl_integration_workspace_alloc(cosmo->gsl_params.N_ITERATION);
-  for(ith=0;ith<n_theta;ith++) {
-    cp->th=theta[ith]*M_PI/180;
-    F.function=&corr_bessel_integrand;
-    F.params=cp;
-    //TODO: Split into intervals between first bessel zeros before integrating
-    //This will help both speed and accuracy of the integral.
-    gslstatus = gsl_integration_qag(&F, 0, cosmo->spline_params.ELL_MAX_CORR, 0,
-                                    cosmo->gsl_params.INTEGRATION_EPSREL, cosmo->gsl_params.N_ITERATION,
-                                    cosmo->gsl_params.INTEGRATION_GAUSS_KRONROD_POINTS,
-                                    w, &result, &eresult);
-    if(gslstatus != GSL_SUCCESS) {
-      ccl_raise_gsl_warning(gslstatus, "ccl_correlation.c: ccl_tracer_corr_bessel():");
-      *status |= gslstatus;
+  gsl_integration_workspace *w = NULL;
+  int local_status;
+
+  #pragma omp parallel default(none) \
+                       shared(cosmo, status, wtheta, n_ell, ell, cls, \
+                              corr_type, cl_spl, theta, n_theta) \
+                       private(w, F, result, eresult, local_status, ith, \
+                               gslstatus, cp)
+  {
+    local_status = *status;
+
+    cp.nell = n_ell;
+    cp.ell0 = ell[0];
+    cp.ellf = ell[n_ell-1];
+    cp.cl0 = cls[0];
+    cp.clf = cls[n_ell-1];
+    switch(corr_type) {
+      case CCL_CORR_GG:
+        cp.i_bessel = 0;
+        break;
+      case CCL_CORR_GL:
+        cp.i_bessel = 2;
+        break;
+      case CCL_CORR_LP:
+        cp.i_bessel = 0;
+        break;
+      case CCL_CORR_LM:
+        cp.i_bessel = 4;
+        break;
+    }
+
+    if (cls[0]*cls[1] <= 0)
+      cp.extrapol_0 = 0;
+    else {
+      cp.extrapol_0 = 1;
+      cp.tilt0 = log10(cls[1]/cls[0])/log10(ell[1]/ell[0]);
+    }
+
+    if (cls[n_ell-2]*cls[n_ell-1] <= 0)
+      cp.extrapol_f = 0;
+    else {
+      cp.extrapol_f = 1;
+      cp.tiltf = log10(cls[n_ell-1]/cls[n_ell-2])/log10(ell[n_ell-1]/ell[n_ell-2]);
+    }
+    cp.cl_spl = cl_spl;
+
+    w = gsl_integration_workspace_alloc(cosmo->gsl_params.N_ITERATION);
+
+    if (w == NULL) {
+      local_status = CCL_ERROR_MEMORY;
+    }
+    F.function = &corr_bessel_integrand;
+    F.params = &cp;
+
+    #pragma omp for
```

Comment:
same comment about `schedule(dynamic)` here. I bet this actually matters for the bessel integrator, since for larger thetas the bessel functions oscillate faster. This may be a moot point though, since I doubt we'll use the bessel integrator in practice.